src/Xamarin.Android.Build.Tasks/Utilities/NdkTools/WithClang.cs

```diff
@@ -0,0 +1,128 @@
+using System;
+using System.IO;
+
+using Microsoft.Build.Utilities;
+using Xamarin.Android.Tools;
+
+namespace Xamarin.Android.Tasks
+{
+	abstract class NdkToolsWithClang : NdkTools
+	{
+		protected string UnifiedHeadersDirPath { get; }
+
+		// See NdkToolsWithClangWithPlatforms.ctor for explanation
+		protected bool NeedClangWorkaround { get; set; }
+
+		protected NdkToolsWithClang (string androidNdkPath, NdkVersion version, TaskLoggingHelper? log)
+			: base (androidNdkPath, version, log)
+		{
+			NdkToolNames[NdkToolKind.CompilerC] = "clang";
+			NdkToolNames[NdkToolKind.CompilerCPlusPlus] = "clang++";
+			UnifiedHeadersDirPath = GetUnifiedHeadersDirPath (androidNdkPath);
+			UsesClang = true;
+		}
+
+		public override bool ValidateNdkPlatform (Action<string> logMessage, Action<string, string> logError, AndroidTargetArch arch, bool enableLLVM)
+		{
+			// Check that we have a compatible NDK version for the targeted ABIs.
+			if (Version.Main.Major < 19) {
+				logMessage (
+					"The detected Android NDK version is incompatible with this version of Xamarin.Android, " +
+					"please upgrade to NDK r19 or newer.");
+			}
+
+			return true;
+		}
+
+		public override int GetMinimumApiLevelFor (AndroidTargetArch arch)
+		{
+			int minValue = 0;
+			string archName = GetPlatformArch (arch);
+			if (!XABuildConfig.ArchAPILevels.TryGetValue (archName, out minValue))
+				throw new InvalidOperationException ($"Unable to determine minimum API level for architecture {arch}");
+
+			return minValue;
+		}
+
+		public override string GetToolPath (NdkToolKind kind, AndroidTargetArch arch, int apiLevel)
+		{
+			string toolName = GetToolName (kind);
+
+			if (kind == NdkToolKind.CompilerC || kind == NdkToolKind.CompilerCPlusPlus) {
+				if (!NeedClangWorkaround) {
+					// See NdkToolsWithClangWithPlatforms.ctor for explanation
+					toolName = $"{GetCompilerTriple (arch)}{apiLevel}-{toolName}";
+				}
+			} else {
+				toolName = $"{GetArchTriple (arch)}-{toolName}";
+			}
+
+			return MakeToolPath (toolName);
+		}
+
+		public override string GetToolPath (string name, AndroidTargetArch arch, int apiLevel)
+		{
+			return MakeToolPath ($"{GetArchTriple (arch)}-{name}");
+		}
+
+		public override string GetClangDeviceLibraryPath ()
+		{
+			string toolchainDir = GetToolchainDir ();
+			string clangBaseDir = Path.Combine (toolchainDir, "lib64", "clang");
+
+			if (!Directory.Exists (clangBaseDir)) {
+				throw new InvalidOperationException ($"Clang toolchain directory '{clangBaseDir}' not found");
+			}
+
+			// There should be just one subdir - clang version - but it's better to be safe than sorry...
+			foreach (string dir in Directory.EnumerateDirectories (clangBaseDir)) {
+				if (dir[0] == '.') {
```

Comment:
Is there a reason we need to explicitly skip over hidden dirs?

Reply:
It's not about hidden directories but `.` and `..` entries. We don't want to travel up, as we know there's nothing there for us to find, and we do know that the current dir also doesn't have what we want - this way we save creation of two strings and at least two `stat` calls. Skipping of the hidden directories is just a bonus (likely irrelevant as it's unlikely any will be found there).

Reply:
`Directory.EnumerateDirectories()` doesn't return `.` or `..`, at least not for me:

```
System.IO.Directory.EnumerateDirectories("/tmp");
// { "/tmp/powerlog", "/tmp/com.apple.launchd.FsyjqY8Rry", "/tmp/.dotnet" }
```

Reply:
Mono must have changed the behavior since that code was implemented, it no longer returns the `.` or `..` entries